### ISSUE TITLE
The active resolution cannot be retrieved for the lowest resolution on the G500s

### ISSUE THREAD

**Issue by Susurrus** [libratbag/libratbag#24]
The active resolution can only be retrieved for the middle and upper DPIs for my G500s, when I have it set to the lowest (so there's only 1 LED on indicating the DPI setting) then it fails to report the active resolution.

So this is me polling the active resolution starting at the highest DPI (all 3 LED indicators are on on the mouse) and then decreasing 1 level before running the command again and then going back up:

```
 ~  Projects  libratbag  sudo tools/ratbag-command resolution active get /dev/input/event5
2
 ~  Projects  libratbag  sudo tools/ratbag-command resolution active get /dev/input/event5
1
 ~  Projects  libratbag  sudo tools/ratbag-command resolution active get /dev/input/event5
Error: Failed to retrieve the active resolution
 ~  Projects  libratbag  sudo tools/ratbag-command resolution active get /dev/input/event5
1
 ~  Projects  libratbag  sudo tools/ratbag-command resolution active get /dev/input/event5
2

```

I tried doing a little debugging, and it looks like the problem is pretty deep. If I use gdb to look at `ratbag_profile_get_resolution()`, the first profile has an `is_active` member of false. So actually retrieving the data from the mouse is incorrect it looks like.

**Comment by cvuchener** [comment]
I do not see `profile->index` being set in `hidpp10drv_read_profile` nor `hidpp10drv_read_profile`, maybe that the reason `is_active` is false. Anyway the support for multi-profile for the G500s is buggy. I think it should only have one profile like Logitech intended, but I am waiting to know more about multi-profile mice to decide.

**Comment by Susurrus** [comment]
Let me know if there's anything I can test for you.

**Comment by cvuchener** [comment]
I have a G500s myself and I am out of idea to test. You can have a look at https://github.com/cvuchener/g500, and tell me if something is wrong, or if you have any idea to complete the doc. The repository contains documentation and low-level tools for tinkering. The only difference I know between the G500 and the G500s is the resolution. Expect g500-parse-profile to show wrong values for that.

I understand that @phomes is working on libratbag's hidpp10 driver code. Maybe he will want a tester.

**Comment by phomes** [comment]
Does the mouse go through the fallback code here:
https://github.com/libratbag/libratbag/blob/master/src/driver-hidpp10.c#L235

**Comment by whot** [comment]
my guess is that it's going through the wrong profile. There were a couple of magic numbers in the hidpp10 code that made it work with my g500s here but if your profile is on a different page then things will just break. There are other pages that may look like valid profiles but aren't.
My profile was on page 5, iirc @cvuchener was on page 2, but I also had valid-looking profiles in pages 3 and 4. So I guess what's happening here is simply that we're picking a wrong profile page.

https://github.com/whot/libratbag/tree/wip/hidpp10-profile-directory should get around this, though it's still WIP.

**Comment by cvuchener** [comment]
> My profile was on page 5, iirc @cvuchener was on page 2, but I also had valid-looking profiles in pages 3 and 4. So I guess what's happening here is simply that we're picking a wrong profile page.

Page 2 on my G500, but page 5 on my G500s. The profile directory is coherent with that. Everything should be fine now.

Yes there are other profile in the memory even though the G500 does not support using multiple profiles. The G500 even has read-only profile at the end of the memory (page 17 and 18, not present on the G500s). Maybe they come from another mouse, maybe Logitech had some plans for them but did not finish it (maybe they wanted to leave some hints about macros to me :)).

**Comment by bentiss** [comment]
Just adding a few data here:

- http://support.logitech.com/en_us/article/14180?product=a0qi00000069ui8AAA says that the G500 has only one on-board profile.
- http://support.logitech.com/en_us/article/18292?product=a0qi00000069uzfAAA says that the G700 has five
- the support pages for the G500s seems to indicate too that the G500s has only one onboard memory profile.

From my observations on the G502 (we still do not have access to the documentation), the profile feature also has memory access and a similar directory (not encoded the exact same way).
The directory on the G502 is good enough to tell us how many profiles are supported and which are enabled/disabled (we have something like `00 01 01 00 00 02 00 00 00 03 01 00 ff ff 00 00` where profile 2 is disabled here).

The only concern I have is that @phomes observed in issue #26 that disabling a profile in the G9 removed it entirely from the list which means we can not count the available profiles with the directory only.

Last (and this is only from observations, not the official Logitech words), the G500 and G700 are the same hardware generation, and I wouldn't be surprised if the firmware is the same except few limitations. It costs less to certify only one firmware and have just a flag to enable/disable features depending on the current hardware. So that's why I think the G500 exports the same memory layout as the G700 but only has one available profile.

**Comment by cvuchener** [comment]
> Last (and this is only from observations, not the official Logitech words), the G500 and G700 are the same hardware generation, and I wouldn't be surprised if the firmware is the same except few limitations. It costs less to certify only one firmware and have just a flag to enable/disable features depending on the current hardware. So that's why I think the G500 exports the same memory layout as the G700 but only has one available profile.

I thought that too and the thirteen buttons in the profile are a hint. But the G700s use a different profile format, I am not so sure any more.

**Comment by bentiss** [comment]
@Susurrus I think the issue is fixed with the push I just made on master. Do you mind testing it again?

Please note that you need to re-install the hwdb file and unplug the device for it to work with the new code.

**Comment by phomes** [comment]
I see something possibly related to this. Any call to read the current resolution fails with "Error: Failed to retrieve the active resolution".

No matter what profile or dpi setting on the mouse hidpp10_get_current_resolution (0x63) just returns 2.

This is for the G9. One thing to note about this mouse is that it has only one setting for dpi, while the others have separate x/y.

**Comment by phomes** [comment]
The issues with the G9 are likely not related. I have a fix specific for that model in #67

**Comment by whot** [comment]
Tested on my G500s here on top of 7778da715eb090fa770bfda96617377522ca7815 and this works now, closing.
